<launch>

    <!-- Arguments -->
        <arg name="world" default="two"/>

        <!-- Gazebo-->
        <arg name="gazebo_gui" default="true"/>
        <arg name="world_file" default="$(find kobuki_navigation)/worlds/$(arg world).world"/>

        <!-- AMCL -->
        <arg name="initial_pose_x"  default="0.0"/>
        <arg name="initial_pose_y"  default="0.0"/>
        <arg name="initial_pose_a"  default="0.0"/>
        <arg name="use_map_topic"   default="false"/>

        <!-- RTAB-->
        <arg name="database_path"     default="rtabmap.db"/>
        <arg name="sw_registered"     default="false"/>
        <arg name="args"  default="--delete_db_on_start"/>
        <arg name="rgb_topic"   default="/camera/rgb/image_raw"/>
        <arg name="depth_topic" default="/camera/depth/image_raw"/>
        <arg name="camera_info_topic" default="/camera/rgb/camera_info"/>
        <arg name="wait_for_transform"  default="0.2"/>


        <!-- Move Base -->
        <arg name="odom_frame_id"   default="odom"/>
        <arg name="base_frame_id"   default="base_footprint"/>
        <arg name="global_frame_id" default="map"/>
        <arg name="odom_topic" default="odom" />
        <arg name="laser_topic" default="scan" />
    
    <param name="use_sim_time" value="true"/>

    <!-- Gazebo -->
    <include file="$(find kobuki_navigation)/launch/includes/gazebo.launch">
        <arg name="gui" value="$(arg gazebo_gui)" />
        <arg name="world_file" value="$(arg world_file)" />
    </include>

    <!-- Map Server for Offline SLAM -->
    <!-- <arg name="map_file" default="$(find kobuki_navigation)/maps/one_one.yaml"/>
    <node name="map_server" pkg="map_server" type="map_server" args="$(arg map_file)" /> -->


    <!-- AMCL (Only for localisation - with static map only) -->
    <!-- <include file="$(find kobuki_navigation)/launch/includes/amcl.launch">
        <arg name="initial_pose_x" value="$(arg initial_pose_x)"/>
        <arg name="initial_pose_y" value="$(arg initial_pose_y)"/>
        <arg name="initial_pose_a" value="$(arg initial_pose_a)"/>
        <arg name="use_map_topic" value="$(arg use_map_topic)"/>

        <arg name="scan_topic" value="$(arg laser_topic)"/>
        <arg name="odom_frame_id" value="$(arg odom_frame_id)"/>
        <arg name="base_frame_id" value="$(arg base_frame_id)"/>
        <arg name="global_frame_id" value="$(arg global_frame_id)"/>
    </include> -->


    <!-- SLAM - Gmapping -->
    <!-- <include file="$(find kobuki_navigation)/launch/includes/gmapping.launch">
        <arg name="scan_topic" value="$(arg laser_topic)"/>
        <arg name="base_frame" value="$(arg base_frame_id)"/>
        <arg name="odom_frame" value="$(arg odom_topic)"/>
    </include> -->


    <!-- SLAM - RTAB 4 -->
    <include file="$(find kobuki_navigation)/launch/includes/rtab.launch">
        <arg name="database_path"       value="$(arg database_path)"/>
        <arg name="sw_registered"       value="$(arg sw_registered)"/>
        <arg name="args"                value="$(arg args)"/>
        <arg name="rgb_topic"           value="$(arg rgb_topic)"/>
        <arg name="depth_topic"         value="$(arg depth_topic)"/>
        <arg name="camera_info_topic"   value="$(arg camera_info_topic)"/>
        <arg name="wait_for_transform"  value="$(arg wait_for_transform)"/>
    </include>


    <!-- move_base -->
    <include file="$(find kobuki_navigation)/launch/includes/move_base.launch">
        <arg name="odom_frame_id" value="$(arg odom_frame_id)"/>
        <arg name="base_frame_id" value="$(arg base_frame_id)"/>
        <arg name="global_frame_id" value="$(arg global_frame_id)"/>
        <arg name="odom_topic" value="$(arg odom_topic)"/>
        <arg name="laser_topic" value="$(arg laser_topic)"/>
    </include>


    <!-- explore_lite (frontier based exploration) -->
    <node pkg="explore_lite" type="explore" respawn="false" name="explore">
        <param name="robot_base_frame" value="base_link"/>
        <param name="costmap_topic" value="map"/>
        <param name="costmap_updates_topic" value="map_updates"/>
        <param name="visualize" value="true"/>
        <param name="planner_frequency" value="0.33"/>
        <param name="progress_timeout" value="30.0"/>
        <param name="potential_scale" value="3.0"/>
        <param name="orientation_scale" value="0.0"/>
        <param name="gain_scale" value="1.0"/>
        <param name="transform_tolerance" value="0.3"/>
        <param name="min_frontier_size" value="0.75"/>
    </node>


    <!-- RViz -->
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find kobuki_navigation)/rviz/custom.rviz"/>

    <!-- Python Nodes -->
    <node pkg="kobuki_navigation" type="image_processing.py" name="image_processing_node" output="screen" />
    <node pkg="kobuki_navigation" type="goal_eval.py" name="goal_eval_node" output="screen" />
    <node pkg="kobuki_navigation" type="manoeuvre.py" name="manoeuvre_node" output="screen" />


</launch>
